### PR TITLE
Feature renormalized photoelectric cross sections and ICRU90 recommended values for key data

### PR DESCRIPTION
**Pull request by mainegra**
This is a major PR introducing several important features:

- Possibility to simulate the photoelectric effect (PE) using Sabbatucci and Salvat's re-normalized cross sections. These are numerically similar to Scofield's 1973 renormalized values (within 0.05%) except that this compilation uses additional energy points around the ionization edges for accurate linear log-log interpolation.
- Ability to generate stopping powers using ICRU90 recommended I values for water and graphite. 
- The choice of atomic relaxation model is now requested via input file rather than via the macro `$EADL_RELAX` which is now deprecated. 
- For consistency with the incoherent scattering default of including binding effects, coherent scattering is now ON by default. Although these effects are negligible in the MV energy range, they have a small impact in simulation time. For instance for a ion chamber simulation in a water phantom irradiated with a Co-60 photon beam the time penalty is about 6% while for a 6 MV beam it is 4%. If this is significant (very long production runs) then the user may turn OFF Rayleigh scattering and use the Klein-Nishina model for incoherent scattering.
- EGSnrc aborts if the user requests bound Compton scattering or electron impact ionization without atomic relaxations. In the past, atomic relaxations were turned ON automatically and a warning issued. However relaxations were only turned ON for the specific interaction and not for the photoelectric effect. Now users are forced to turn ON relaxation in their input files.

To use the features in this PR before they are merged with the `develop` branch users can get the branch `feature-shellwise-pe-xsections` and create a user configuration for it. See bottom of this PR for more details on this.

To use the re-normalized cross sections one will have to set the input key `Photon cross sections` in the `MC transport parameter` input block to one of the following values:
```
:start MC transport parameter:
  ...
  Photon cross sections = mcdf-xcom # or mcdf-epdl
  Atomic relxations = On # Off, EADL, simple
  ...
:stop MC transport parameter:

```
These options allow the use of the re-normalized PE cross sections while using the rest of the photon cross sections from either the XCOM or EPDL libraries. Note above that EADL relaxation is turned ON by using the values `On` or `EADL`. If the input is `simple`, then the original relaxation implementation is used. However, when `mcdf-xcom` or `mcdf-epdl `is used, no `simple` atomic relaxation is allowed.

The re-normalized cross sections are available for all atomic sub-shells up to N7 for elements from Z=1 to Z=99 for energies from the ionization threshold up to 1 GeV. However only sub-shells above 1 keV are included in EGSnrc. Taking advantage of the availability of sub-shell PE cross sections a new routine is implemented that allows sampling the photoelectric interaction from all energetically possible sub-shells. Since an adaptive energy grid is used near absorption edges, to ensure accurate interpolation, a binary search algorithm is used in the new photoelectric routine to find the energy interval the photon energy falls in. However, the total PE cross section used for sampling the interaction uses the same equidistant grid as the rest of the photon cross sections for fast interpolation. The time penalty of the new algorithm is about 6% for a BEAMnrc simulation of a 30 kV x-ray tube. However for an `egs_fac` simulation of HVL attenuation measurements with a free-air chamber (FAC) for 4 attenuation filters including the BEAMnrc x-ray tube simulation no time difference is observed. This can be attributed to the fact that more very low-energy particles are produced in the BEAMnrc simulation that don't make it to the FAC.

To create material data for graphite or water according to ICRU90 recommendations one will have to make use of the following density correction files: 

```
$HEN_HOUSE/pegs4/density_corrections/elements/carbon_graphite_ICRU90_1.7g_cm3.density 
$HEN_HOUSE/pegs4/density_corrections/elements/carbon_graphite_ICRU90_2.0g_cm3.density
$HEN_HOUSE/pegs4/density_corrections/compounds/water_ICRU90.density                   
```
The easiest way is to take advantage of the pegsless option which generates the data _on-the-fly_ by adding the following input block to the input file:
```
:start media definition:
    ### energy transport cutoffs
    ae = 0.512
    ue = 2.511
    ap = 0.001
    up = 2.00
    :start water_icru90:
        density correction file     = water_icru90
        bremsstrahlung correction   = NRC
    :stop water_icru90:
    :start c170_icru90:
      density correction file     =  carbon_graphite_ICRU90_1.7g_cm3
        bremsstrahlung correction   = NRC
    :stop c170_icru90:
    :start c200_icru90:
      density correction file     =  carbon_graphite_ICRU90_2.0g_cm3
        bremsstrahlung correction   = NRC
    :stop c200_icru90:
:stop media definition:

```
One can of course create a pegs4 data file for these media making sure to use the proper density correction files.

Several checks have been made to ensure that the implementation produces correct values. The graph below shows the ratio of air mass-energy absorption coefficients calculated with the EGSnrc app `g.mortran` either with the original XCOM PE cross sections or with the re-normalized cross sections of Sabbatucci and Salvat to the Penelope values taken from the ICRU90 report. The first shows similar differences as reported in ICRU90 report while the second demonstrates excellent agreement between both codes except at 3.263 keV (Ar K shell binding energy in Penelope) because of the small differences in binding energies between the codes (EGSnrc uses the XCOM value of 3.2029 keV).
muen_rho_air_egsnrc_vs_pen_1

To make use of these features one can [clone the git repository](https://github.com/nrc-cnrc/EGSnrc/wiki/install-EGSnrc-on-linux#2-download-egsnrc) for EGSnrc and then all branches are available to the user. To switch to the branch `feature-shellwise-pe-xsections` one needs to execute

`git checkout feature-shellwise-pe-xsections `

on the command line in the folder containing the git repository. Once on that branch the user can create a configuration either for [Linux](https://github.com/nrc-cnrc/EGSnrc/wiki/install-EGSnrc-on-linux#3-configure-egsnrc) or [Windows](https://github.com/nrc-cnrc/EGSnrc/wiki/Install-EGSnrc-on-Windows#3-configure-egsnrc).

Alternatively to using git, one can download this branch directly as a zip archive and then configure it as a separate install using one of the methods mentioned in the previous paragraph. Go to the [EGSnrc github page of this branch](https://github.com/nrc-cnrc/EGSnrc/tree/feature-shellwise-pe-xsections) and download it by clicking the green button on the right `Clone or download` and click on the `Download ZIP` link.

**Details to take into consideration when using this branch and dosxyznrc:**
- If one only chooses to create a new configuration for this branch, the user area `$EGS_HOME` remains untouched.
- As a consequence `dosxyznrc.mortran` is not updated, hence its compilation will fail during the execution of `EGSnrc-configure-linux` as this file was modified for this branch.
- The solution is to update `dosxyznrc.mortran` from $HEN_HOUSE/user_codes/dosxyznrc/
- Setup the environment variable `$EGS_CONFIG` in .bashrc to point to the new configuration and start a new shell console.

### CODE REVIEW

File: HEN_HOUSE/omega/progs/gui/beamnrc/beamnrc_params.tcl [hunk 1]
```diff
@@ -1649,33 +1653,43 @@ set names(iedgfl) "Atomic relaxations"
...
-- Shell vacancies created in photo-absorption events\
+  cross sections\
+- Shell vacancies created in these events\
   are relaxed via emission of fluorescent X-Rays,\
   Auger and Koster-Cronig electrons.\
   Make sure to turn this option on for low energy\
   applications.\
 
+Selecting On or EADl is equivalent. A full atomic relaxation cascade\
+is simmulated using EADL transition probabilities.
```
Comment (rtownson) [inline HEN_HOUSE/omega/progs/gui/beamnrc/beamnrc_params.tcl]: Typo in `simulated`

File: HEN_HOUSE/omega/progs/gui/dosxyznrc/xyznrc_parameters.tcl [hunk 2]
```diff
@@ -1212,23 +1216,34 @@ set options(iraylr,1) "On"
...
+  cross sections
+- Shell vacancies created in these events\
   are relaxed via emission of fluorescent X-Rays,\
   Auger and Koster-Cronig electrons.\
   Make sure to turn this option on for low energy\
-  applications.\
+  applications.
+
+Selecting On or EADl is equivalent. A full atomic relaxation cascade\
+is simmulated using EADL transition probabilities.
```
Comment (rtownson) [inline HEN_HOUSE/omega/progs/gui/dosxyznrc/xyznrc_parameters.tcl]: Typo

File: HEN_HOUSE/src/egs_utilities.mortran [hunk 3]
```diff
@@ -1629,6 +1630,39 @@ IF( which = 1 ) [
 ]
 return; end;
 
+/* Print binding energies: In the case of the EPDL library
+   only energies above 1 keV are output for elements not part
+   of the current simlation. This is due to the fact that energies 
+   below 1 keV are taken from the relaxation database only for 
+   elements requested in the input when EPDL is used. This is not
+   and issue when using XCOM.
```
Comment (rtownson) [inline HEN_HOUSE/src/egs_utilities.mortran]: Typo: `and` should be `an`

File: HEN_HOUSE/src/get_inputs.mortran [hunk 4]
```diff
@@ -1690,6 +1713,28 @@ write(i_errors,*)
 ' ******************** end input transport parameter *********************** ';
 write(i_errors,*);
 
+"Check if EADL relaxation requested. Note that original relaxation"
+"algorithm using <M> and <N> is only turned ON for all regions."
+"Moved passed the $TURN-ON/OFF-IN-REGIONS statement to catch the"
```
Comment (rtownson) [inline HEN_HOUSE/src/get_inputs.mortran]: Typo: `passed` should be `past`

Comment (mainegra) [inline HEN_HOUSE/src/get_inputs.mortran]: Thanks a bunch @rtownson for reviewing! Updating the documentation is the next step.

**Comment by ftessier** [comment]
Ernesto, do you want to fix the noted typos? You can add a commit at the end; or else if you prefer you can use `git rebase -i af4a48a` and change `pick` to `edit` for the commits where the typos appeared. Then git replays the commits and stop for the ones you wanted to modify. Edit the files, add and commit, and keep going with `git rebase --continue`. Loads of fun!

**Comment by mainegra** [comment]
Thanks @blakewalters  for reviewing!

**Comment by mainegra** [comment]
@ftessier I will correct the typos found by @rtownson ! Seems easier! :-)

**Comment by mainegra** [comment]
Done!

**Comment by ftessier** [comment]
@mainegra can you `rebase -i origin/develop`? There is a conflict which I am not sure how to resolve, viz. should the explanation in the comment remain? I presume not, just want to make sure. Can we remove the commented out IF block?
```
<<<<<<< f0c54d1ebe26c5a1acd464440846f458990bc101
    "Assign this energy deposition to an electron."
    "Note that this should NOT be treated as though it came from a photon,"
    "even if a photon initiated the relaxations. Rather, energy remaining"
    "in vacancies should be given to an electron or absorbed locally."
    $AUSCALL($SELECTRONA); "call ausgab with iarg=34"
||||||| merged common ancestors
    IF( relax_initiator_q = 0 ) [
        "particle initiating relaxations was photon"
        $AUSCALL($SPHOTONA);  "call ausgab with iarg=33"
    ] ELSE [
        "particle initiating relaxations was electron"
        $AUSCALL($SELECTRONA); "call ausgab with iarg=34"
    ]
=======
    /*IF( relax_initiator_q = 0 ) [
        "particle initiating relaxations was photon"
        $AUSCALL($SPHOTONA);  "call ausgab with iarg=33"
    ] ELSE [
        "particle initiating relaxations was electron"
        $AUSCALL($SELECTRONA); "call ausgab with iarg=34"
    ]*/
    $AUSCALL($SELECTRONA); "call ausgab with iarg=34"
>>>>>>> Implement shellwise photoelectric interaction
```

**Comment by ftessier** [comment]
Rebased on tip of `develop` to fix the merge conflict.

**Comment by rtownson** [comment]
Now that you've rebased you can remove the commented out block Fred referenced. `relax_initiator_q` is no longer used, just the `$AUSCALL($SELECTRONA);` line should remain.

**Comment by ftessier** [comment]
I though I removed it; is it still there?

**Comment by ftessier** [comment]
Updated commit history (reorder, reword, edit, squash, etc.), adjusted contributors in `egs_functions.cpp` and `egs_functions.h`, removed EOL white space and ran astyle in egs++ directory. The few code changes can be reviewed with `git diff a5aa2b1`.

**Comment by ftessier** [comment]
Just squashed the EDKnrc update for `ibsearch` into commit 26156f5.

**Pull request merged**